params.go

```diff
@@ -347,12 +347,21 @@ func BuildQueryString(opts interface{}) (*url.URL, error) {
 								params.Add(tags[0], v.Index(i).String())
 							}
 						}
+					case reflect.Map:
+						if v.Type().Key().Kind() == reflect.String && v.Type().Elem().Kind() == reflect.String {
+							var s []string
+							for _, k := range v.MapKeys() {
+								value := v.MapIndex(k).String()
+								s = append(s, fmt.Sprintf("'%s':'%s'", k.String(), value))
+							}
+							params.Add(tags[0], fmt.Sprintf("{%s}", strings.Join(s, ", ")))
+						}
 					}
 				} else {
 					// Otherwise, the field is not set.
 					if len(tags) == 2 && tags[1] == "required" {
 						// And the field is required. Return an error.
-						return nil, fmt.Errorf("Required query parameter [%s] not set.", f.Name)
+						return &url.URL{}, fmt.Errorf("Required query parameter [%s] not set.", f.Name)
```

Comment:
This should be `nil` instead of `&url.URL{}`.

Reply:
It seems that other methods which invoke BuildQueryString do not check the error first, but call String() of first returned value directly, which will leads to a run time error. see
https://github.com/gophercloud/gophercloud/blob/master/openstack/blockstorage/v2/volumes/requests.go#L99

Reply:
https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/servers/requests.go#L63

Reply:
ah, yes, that would definitely case a problem. Good catch.